File: apps/mobile/android/app/src/main/java/com/visvine/mobile/data/realtime/MessageStream.kt
package com.visvine.mobile.data.realtime

import com.visvine.mobile.core.AppConfig
import com.visvine.mobile.data.local.TokenStore
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonPrimitive
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.sse.EventSource
import okhttp3.sse.EventSourceListener
import okhttp3.sse.EventSources
import javax.inject.Inject
import javax.inject.Singleton

/**
 * One realtime event off `/api/messages/stream`. We only surface the discriminant
 * `type` and the optional `conversationId`; screens re-fetch on relevant events
 * rather than patching state from the payload. This is foreground delivery only —
 * background delivery needs push, which is not wired up yet.
 */
data class RealtimeEvent(
    val type: String,
    val conversationId: String?,
)

/**
 * SSE consumer. The browser EventSource API can't set headers, so — like the
 * iOS plan — we hand-roll over OkHttp with an `Authorization: Bearer` header and
 * surface connection drops so callers can reconnect. A 401 (expired 30-day JWT)
 * arrives as [EventSourceListener.onFailure]; the AuthViewModel treats a failed
 * session refresh as the signal to clear the token.
 */
@Singleton
class MessageStream @Inject constructor(
    private val client: OkHttpClient,
    private val tokenStore: TokenStore,
    private val json: Json,
) {
    fun events(): Flow<RealtimeEvent> = callbackFlow {
        val token = tokenStore.currentToken()
        val request = Request.Builder()
            .url("${AppConfig.apiBaseUrl}/api/messages/stream")
            .header("Accept", "text/event-stream")
            .apply { if (!token.isNullOrBlank()) header("Authorization", "Bearer $token") }
            .build()

        val listener = object : EventSourceListener() {
            override fun onEvent(
                eventSource: EventSource,
                id: String?,
                type: String?,
                data: String,
            ) {
                // Backend sends `data: {json}` lines; keepalive/comment lines are
                // delivered as comments and never reach onEvent.
                runCatching {
                    val obj = json.parseToJsonElement(data) as? JsonObject ?: return
                    val eventType = obj["type"]?.jsonPrimitive?.content ?: return
                    val convoId = obj["conversationId"]?.jsonPrimitive?.content
                    trySend(RealtimeEvent(eventType, convoId))
                }
            }

            override fun onFailure(
                eventSource: EventSource,
                t: Throwable?,
                response: okhttp3.Response?,
            ) {
                close(t)
            }

            override fun onClosed(eventSource: EventSource) {
                close()
            }
        }

        val source = EventSources.createFactory(client).newEventSource(request, listener)
        awaitClose { source.cancel() }
    }
}

File: apps/mobile/ios/Visvine/Realtime/MessageStream.swift
import Foundation

/// One realtime event off `/api/messages/stream`. Screens re-fetch on relevant
/// events rather than patching from the payload.
struct RealtimeEvent {
    let type: String
    let conversationId: String?
}

/// SSE consumer over `URLSession.bytes` so we can set the `Authorization: Bearer`
/// header (the browser EventSource API can't). This is foreground delivery only —
/// background delivery needs push, which is not wired up yet.
/// A 401 (expired 30-day JWT) ends the stream; callers refresh the session.
final class MessageStream {
    static let shared = MessageStream()

    func events() -> AsyncStream<RealtimeEvent> {
        AsyncStream { continuation in
            let task = Task {
                guard let url = URL(string: AppConfig.apiBase + "/api/messages/stream") else {
                    continuation.finish(); return
                }
                var request = URLRequest(url: url)
                request.setValue("text/event-stream", forHTTPHeaderField: "Accept")
                if let token = KeychainTokenStore.shared.token {
                    request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
                }
                do {
                    let (bytes, response) = try await URLSession.shared.bytes(for: request)
                    if let http = response as? HTTPURLResponse, !(200..<300).contains(http.statusCode) {
                        continuation.finish(); return
                    }
                    for try await line in bytes.lines {
                        // SSE: payload lines start with `data:`; `:` comments (keepalive) are skipped.
                        guard line.hasPrefix("data:") else { continue }
                        let payload = line.dropFirst(5).trimmingCharacters(in: .whitespaces)
                        guard !payload.isEmpty,
                              let data = payload.data(using: .utf8),
                              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                              let type = obj["type"] as? String else { continue }
                        continuation.yield(RealtimeEvent(type: type, conversationId: obj["conversationId"] as? String))
                    }
                } catch {
                    // Connection drop / cancellation — caller may reconnect.
                }
                continuation.finish()
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }
}
